File: sdks/python/taskito/mixins/pubsub.py
"""Topic pub/sub: fan-out subscribers (one job each) or log subscribers (pull)."""

from __future__ import annotations

import json
import math
import threading
from collections.abc import Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from taskito.mixins._log_consumer import LogConsumerThread
from taskito.notes import validate_and_encode_notes
from taskito.result import JobResult

if TYPE_CHECKING:
    from taskito._taskito import PyQueue
    from taskito.serializers import Serializer
    from taskito.task import TaskWrapper


@dataclass(frozen=True)
class TopicMessage:
    """One message pulled from a log topic.

    ``id`` is the cursor token to pass to :meth:`ack_topic`. ``args``/``kwargs``
    are the decoded publish payload; ``metadata``/``notes`` are the caller's.
    """

    id: str
    args: tuple[Any, ...]
    kwargs: dict[str, Any]
    metadata: dict[str, Any] | None
    notes: dict[str, Any] | None
    created_at: int


class QueuePubSubMixin:
    """Publish/subscribe over topics.

    A subscriber is a normal task plus a registry row mapping
    ``(topic, subscription_name) -> (task_name, queue)``. ``publish()`` fans a
    message out as one ordinary job per active subscription, so every task
    feature — retries, DLQ, middleware, rate limits — applies per subscriber,
    and a failing subscriber never affects its siblings.

    Deliveries use the queue-level serializer (there is one payload for all
    subscribers); per-task serializer overrides don't apply to topic tasks.
    """

    _inner: PyQueue
    _serializer: Serializer
    _subscription_configs: list[dict[str, Any]]
    _log_consumer_configs: list[dict[str, Any]]

    def subscriber(
        self,
        topic: str,
        name: str | None = None,
        queue: str = "default",
        durable: bool = True,
        **task_kwargs: Any,
    ) -> Callable[[Callable[..., Any]], TaskWrapper]:
        """Decorator registering ``fn`` as an independent subscriber of ``topic``.

        The function becomes a normal task (``task_kwargs`` are forwarded to
        :meth:`task`), and the subscription is written to storage when
        ``run_worker()`` starts — or via :meth:`declare_subscriptions` in a
        producer-only process. ``durable=False`` ties the subscription to one
        worker process: it only registers inside ``run_worker()`` and is
        reaped automatically once its worker stops heartbeating.

        Args:
            topic: Topic to subscribe to.
            name: Stable subscription identity. Defaults to the task name;
                re-registering the same ``(topic, name)`` updates the routing
                target instead of duplicating.
            queue: Queue the subscriber's delivery jobs go to.
            durable: Persist across restarts (default). ``False`` = ephemeral.
            **task_kwargs: Any :meth:`task` option (``max_retries``,
                ``timeout``, ``middleware``, ``idempotent``, ...).
        """

        if "codecs" in task_kwargs:
            raise ValueError(
                "subscriber() does not accept per-task codecs: publish() encodes one "
                "shared payload with the queue serializer, so a per-subscriber codec "
                "chain could never be applied on the producer side"
            )

        def decorator(fn: Callable[..., Any]) -> TaskWrapper:
            wrapper: TaskWrapper = self.task(queue=queue, **task_kwargs)(fn)  # type: ignore[attr-defined]
            config = {
                "topic": topic,
                "subscription_name": name or wrapper.name,
                "task_name": wrapper.name,
                "queue": queue,
                "durable": durable,
            }
            # Replace, don't append: re-decorating the same (topic, name) (module
            # reloads, test fixtures) must not accumulate stale declarations.
            identity = (config["topic"], config["subscription_name"])
            self._subscription_configs[:] = [
                c
                for c in self._subscription_configs
                if (c["topic"], c["subscription_name"]) != identity
            ]
            self._subscription_configs.append(config)
            return wrapper

        return decorator

    def subscribe_log(self, topic: str, name: str) -> None:
        """Register a durable **log** subscription (a named cursor over ``topic``).

        Unlike :meth:`subscriber`, a log subscription has no handler: the topic's
        publishes are stored once each and this consumer pulls them with
        :meth:`read_topic`, advancing its cursor with :meth:`ack_topic`. Writes
        immediately to storage, so a producer must have registered it (or called
        this) before the publishes it wants to see.
        """
        self._inner.register_subscription(
            topic=topic,
            subscription_name=name,
            task_name="",
            queue="default",
            durable=True,
            owner_worker_id=None,
            mode="log",
        )

    def log_consumer(
        self,
        topic: str,
        name: str | None = None,
        *,
        poll_interval: float = 1.0,
        batch_size: int = 100,
        on_error: str = "retry",
    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
        """Decorator: register ``fn`` as a managed consumer of log ``topic``.

        Registers a durable log subscription (like :meth:`subscribe_log`) and,
        when :meth:`run_worker` starts, spawns a daemon thread that pulls
        messages, invokes ``fn(*args, **kwargs)`` per message, and advances the
        cursor — the read/ack loop callers otherwise hand-write. The handler may
        be sync or async; a producer-only process that never runs a worker still
        registers the subscription so its publishes are retained.

        Run a given ``(topic, name)`` consumer in **one** worker process — the
        cursor is a single high-water mark, so parallel workers would double-read.
        Use distinct ``name`` values for independent consumers.

        Args:
            topic: Log topic to consume.
            name: Stable subscription identity. Defaults to the handler name.
            poll_interval: Seconds to wait after an empty poll before re-reading.
            batch_size: Max messages pulled per poll.
            on_error: ``"retry"`` (default) leaves the failed message un-acked so
                the batch re-reads; ``"skip"`` acks past it and continues.
        """
        if on_error not in ("retry", "skip"):
            raise ValueError(f"on_error must be 'retry' or 'skip', got {on_error!r}")
        # A non-finite/non-positive interval spins on empty polls; a non-positive
        # batch size never makes progress.
        if not (poll_interval > 0 and math.isfinite(poll_interval)):
            raise ValueError(
                f"poll_interval must be a finite positive number, got {poll_interval!r}"
            )
        if batch_size <= 0:
            raise ValueError(f"batch_size must be a positive integer, got {batch_size!r}")

        def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
            sub_name = name or fn.__name__
            self.subscribe_log(topic, sub_name)
            config = {
                "topic": topic,
                "name": sub_name,
                "handler": fn,
                "poll_interval": poll_interval,
                "batch_size": batch_size,
                "on_error": on_error,
            }
            # Replace, don't append: re-decorating the same (topic, name) (module
            # reloads, test fixtures) must not spawn duplicate consumer threads.
            identity = (topic, sub_name)
            self._log_consumer_configs[:] = [
                c for c in self._log_consumer_configs if (c["topic"], c["name"]) != identity
            ]
            self._log_consumer_configs.append(config)
            return fn

        return decorator

    def _build_log_consumer_threads(self, stop_event: threading.Event) -> list[LogConsumerThread]:
        """One :class:`LogConsumerThread` per registered consumer, sharing a
        single stop event so ``run_worker`` can drain them all at once."""
        return [
            LogConsumerThread(self, config, stop_event) for config in self._log_consumer_configs
        ]

    def read_topic(self, topic: str, name: str, limit: int = 100) -> list[TopicMessage]:
        """Pull up to ``limit`` messages after a log subscription's cursor.

        Oldest first, exclusive of the cursor. Decodes each payload with the
        queue serializer. Returns an empty list when the consumer is caught up.
        Delivery is at-least-once: process, then :meth:`ack_topic` the last id.
        """
        messages = []
        for msg_id, payload, metadata, notes, created_at in self._inner.read_topic_messages(
            topic, name, limit
        ):
            args, kwargs = self._serializer.loads(payload)
            messages.append(
                TopicMessage(
                    id=msg_id,
                    args=args,
                    kwargs=kwargs,
                    metadata=json.loads(metadata) if metadata is not None else None,
                    notes=json.loads(notes) if notes is not None else None,
                    created_at=created_at,
                )
            )
        return messages

    def ack_topic(self, topic: str, name: str, cursor: str) -> bool:
        """Advance a log subscription's cursor to ``cursor`` (a message id).

        A high-water mark: acking an id acks everything up to and including it.
        Monotonic — acking an older id is a no-op. Returns False if nothing moved.
        """
        return self._inner.ack_topic_cursor(topic, name, cursor)

    def topic_log_stats(self) -> list[dict[str, Any]]:
        """Lag snapshot per log subscription.

        Each entry: ``topic``, ``subscription``, ``cursor`` (``None`` if unread),
        ``lag`` (un-acked messages), ``oldest_unacked_age_ms`` (``None`` if
        caught up).
        """
        return [
            {
                "topic": row[0],
                "subscription": row[1],
                "cursor": row[2],
                "lag": row[3],
                "oldest_unacked_age_ms": row[4],
            }
            for row in self._inner.topic_log_stats()
        ]

    def declare_subscriptions(self) -> None:
        """Write pending durable subscriptions to storage.

        Called automatically at ``run_worker()`` startup. Call it explicitly
        in a producer-only process (one that imports subscriber modules but
        never runs a worker) so ``publish()`` sees the subscriptions.
        Ephemeral subscriptions are skipped — they need an owning worker.
        """
        for config in self._subscription_configs:
            if config["durable"]:
                self._register_subscription(config, owner_worker_id=None)

    def publish(
        self,
        topic: str,
        *args: Any,
        idempotency_key: str | None = None,
        metadata: dict[str, Any] | None = None,
        notes: dict[str, Any] | None = None,
        priority: int | None = None,
        delay_seconds: float | None = None,
        max_retries: int | None = None,
        timeout: int | None = None,
        expires: float | None = None,
        result_ttl: int | None = None,
        **kwargs: Any,
    ) -> list[JobResult]:
        """Publish a message to ``topic``.

        Every active subscription receives an independent job carrying the
        same ``(args, kwargs)`` payload (at-least-once per subscriber).
        Returns one :class:`JobResult` per delivery — empty when the topic
        has no active subscribers, which is a valid pub/sub no-op.

        ``idempotency_key`` dedupes per subscriber: republishing the same key
        yields no new deliveries, and a subscription added later still gets
        its own copy. Each delivery's ``notes`` carry ``topic`` and
        ``subscription`` for filtering.
        """
        payload = self._serializer.dumps((args, kwargs))
        self._check_payload_size(topic, len(payload))  # type: ignore[attr-defined]
        py_jobs = self._inner.publish(
            topic=topic,
            payload=payload,
            idempotency_key=idempotency_key,
            metadata=json.dumps(metadata) if metadata is not None else None,
            notes=validate_and_encode_notes(notes) if notes is not None else None,
            priority=priority,
            delay_seconds=delay_seconds,
            max_retries=max_retries,
            timeout=timeout,
            expires=expires,
            result_ttl=result_ttl,
        )
        return [JobResult(py_job=py_job, queue=self) for py_job in py_jobs]  # type: ignore[arg-type]

    def unsubscribe(self, topic: str, name: str) -> bool:
        """Remove a subscription. Returns False if none matched.

        Also drops any matching local declaration so a later worker start
        doesn't re-register what was just removed.
        """
        self._subscription_configs[:] = [
            c
            for c in self._subscription_configs
            if (c["topic"], c["subscription_name"]) != (topic, name)
        ]
        return self._inner.unsubscribe(topic, name)

    def pause_subscription(self, topic: str, name: str) -> bool:
        """Stop deliveries without unregistering. Returns False if unknown."""
        return self._inner.set_subscription_active(topic, name, False)

    def resume_subscription(self, topic: str, name: str) -> bool:
        """Resume a paused subscription. Returns False if unknown."""
        return self._inner.set_subscription_active(topic, name, True)

    def list_subscriptions(self, topic: str | None = None) -> list[dict[str, Any]]:
        """List subscriptions — all of them, or one topic's active ones."""
        rows = self._inner.list_subscriptions(topic)
        return [
            {
                "topic": row[0],
                "name": row[1],
                "task_name": row[2],
                "queue": row[3],
                "active": row[4],
                "durable": row[5],
            }
            for row in rows
        ]

    def list_topics(self) -> list[str]:
        """Distinct topics that currently have at least one subscription."""
        seen: dict[str, None] = {}
        for row in self._inner.list_subscriptions(None):
            seen.setdefault(row[0], None)
        return list(seen)

    def topic_stats(self, topic: str | None = None) -> list[dict[str, Any]]:
        """Backlog/lag snapshot per subscription, optionally filtered to a topic.

        Each entry: ``topic``, ``subscription``, ``task_name``, ``queue``,
        ``active``, ``durable``, ``pending``, ``running``, ``dead``, and
        ``oldest_pending_age_ms`` (``None`` when the subscription has no pending
        backlog). Computed live off indexed columns — safe to poll.
        """
        stats = [
            {
                "topic": row[0],
                "subscription": row[1],
                "task_name": row[2],
                "queue": row[3],
                "active": row[4],
                "durable": row[5],
                "pending": row[6],
                "running": row[7],
                "dead": row[8],
                "oldest_pending_age_ms": row[9],
            }
            for row in self._inner.topic_backlog_stats()
        ]
        return stats if topic is None else [s for s in stats if s["topic"] == topic]

    def _delivery_task_defaults(self) -> dict[str, tuple[int, int, int]]:
        """Per-task ``(priority, max_retries, timeout_ms)`` from this process's
        task registry. Persisted on the subscription row at registration so a
        producer-only process (which never loads the task) still applies each
        subscriber's own settings. Tasks not registered here get queue defaults."""
        return {
            config.name: (config.priority, config.max_retries, config.timeout * 1000)
            for config in self._task_configs  # type: ignore[attr-defined]
        }

    def _register_subscription(self, config: dict[str, Any], owner_worker_id: str | None) -> None:
        priority, max_retries, timeout_ms = self._delivery_task_defaults().get(
            config["task_name"], (None, None, None)
        )
        self._inner.register_subscription(
            topic=config["topic"],
            subscription_name=config["subscription_name"],
            task_name=config["task_name"],
            queue=config["queue"],
            durable=config["durable"],
            owner_worker_id=owner_worker_id,
            priority=priority,
            max_retries=max_retries,
            timeout_ms=timeout_ms,
        )

    def _declare_worker_subscriptions(self, worker_id: str) -> None:
        """Flush all subscriptions at worker startup, owning the ephemeral ones."""
        for config in self._subscription_configs:
            owner = None if config["durable"] else worker_id
            self._register_subscription(config, owner_worker_id=owner)

File: crates/taskito-python/src/py_queue/pubsub.rs
use pyo3::prelude::*;

use taskito_core::job::now_millis;
use taskito_core::pubsub::{publish_to_topic, DeliveryDefaults, PublishRequest};
use taskito_core::storage::records::NewSubscription;
use taskito_core::storage::Storage;

use super::PyQueue;
use crate::py_job::PyJob;

/// A subscription row surfaced to Python:
/// `(topic, subscription_name, task_name, queue, active, durable)`.
type SubscriptionTuple = (String, String, String, String, bool, bool);

/// Per-subscription backlog snapshot surfaced to Python:
/// `(topic, subscription, task_name, queue, active, durable, pending, running,
/// dead, oldest_pending_age_ms)`.
type SubscriptionBacklogTuple = (
    String,
    String,
    String,
    String,
    bool,
    bool,
    i64,
    i64,
    i64,
    Option<i64>,
);

/// A log message surfaced to Python:
/// `(id, payload, metadata, notes, created_at)`.
type TopicMessageTuple = (String, Vec<u8>, Option<String>, Option<String>, i64);

/// Per-log-subscription lag surfaced to Python:
/// `(topic, subscription, cursor, lag, oldest_unacked_age_ms)`.
type TopicLogStatsTuple = (String, String, Option<String>, i64, Option<i64>);

/// A declared topic surfaced to Python: `(name, mode, retention_ms, created_at)`.
type TopicTuple = (String, String, Option<i64>, i64);

#[pymethods]
impl PyQueue {
    /// Insert or update a topic subscription (idempotent on topic + name).
    ///
    /// `priority`/`max_retries`/`timeout_ms` (already in milliseconds) persist
    /// the subscriber task's own delivery settings so a producer-only process
    /// can apply them without loading the task.
    #[pyo3(signature = (topic, subscription_name, task_name, queue="default", durable=true, owner_worker_id=None, priority=None, max_retries=None, timeout_ms=None, mode="fanout"))]
    #[allow(clippy::too_many_arguments)]
    pub fn register_subscription(
        &self,
        topic: &str,
        subscription_name: &str,
        task_name: &str,
        queue: &str,
        durable: bool,
        owner_worker_id: Option<&str>,
        priority: Option<i32>,
        max_retries: Option<i32>,
        timeout_ms: Option<i64>,
        mode: &str,
    ) -> PyResult<()> {
        // An unowned ephemeral row could never be reaped (cleanup keys off
        // live worker ids), so it would stay active forever.
        if !durable && owner_worker_id.is_none() {
            return Err(pyo3::exceptions::PyValueError::new_err(
                "an ephemeral subscription (durable=false) requires owner_worker_id",
            ));
        }
        let row = NewSubscription {
            topic: topic.to_string(),
            subscription_name: subscription_name.to_string(),
            task_name: task_name.to_string(),
            queue: queue.to_string(),
            active: true,
            durable,
            owner_worker_id: owner_worker_id.map(str::to_string),
            created_at: now_millis(),
            priority,
            max_retries,
            timeout_ms,
            mode: mode.to_string(),
        };
        self.storage
            .register_subscription(&row)
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// List subscriptions — all of them, or only a topic's active ones.
    #[pyo3(signature = (topic=None))]
    pub fn list_subscriptions(&self, topic: Option<&str>) -> PyResult<Vec<SubscriptionTuple>> {
        let rows = match topic {
            Some(t) => self.storage.list_subscriptions_for_topic(t),
            None => self.storage.list_subscriptions(),
        }
        .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?;
        Ok(rows
            .into_iter()
            .map(|row| {
                (
                    row.topic,
                    row.subscription_name,
                    row.task_name,
                    row.queue,
                    row.active,
                    row.durable,
                )
            })
            .collect())
    }

    /// Remove a subscription. Returns false if none matched.
    pub fn unsubscribe(&self, topic: &str, subscription_name: &str) -> PyResult<bool> {
        self.storage
            .unsubscribe(topic, subscription_name)
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Pause (false) or resume (true) a subscription without unregistering it.
    pub fn set_subscription_active(
        &self,
        topic: &str,
        subscription_name: &str,
        active: bool,
    ) -> PyResult<bool> {
        self.storage
            .set_subscription_active(topic, subscription_name, active)
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Backlog/lag snapshot per subscription:
    /// `(topic, subscription, task_name, queue, active, durable, pending,
    /// running, dead, oldest_pending_age_ms)`.
    pub fn topic_backlog_stats(&self, py: Python<'_>) -> PyResult<Vec<SubscriptionBacklogTuple>> {
        let storage = &self.storage;
        // Release the GIL: aggregation groups over every pub/sub delivery.
        py.detach(|| storage.topic_backlog_stats())
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?
            .into_iter()
            .map(|s| {
                Ok((
                    s.topic,
                    s.subscription_name,
                    s.task_name,
                    s.queue,
                    s.active,
                    s.durable,
                    s.pending,
                    s.running,
                    s.dead,
                    s.oldest_pending_age_ms,
                ))
            })
            .collect()
    }

    /// Messages after a log subscription's cursor, oldest first, `<= limit`:
    /// `(id, payload, metadata, notes, created_at)`. Cursor resolved server-side.
    pub fn read_topic_messages(
        &self,
        py: Python<'_>,
        topic: &str,
        subscription_name: &str,
        limit: i64,
    ) -> PyResult<Vec<TopicMessageTuple>> {
        let storage = &self.storage;
        Ok(py
            .detach(|| storage.read_topic_messages(topic, subscription_name, limit))
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?
            .into_iter()
            .map(|m| (m.id, m.payload, m.metadata, m.notes, m.created_at))
            .collect())
    }

    /// Advance a log subscription's cursor to `cursor` (monotonic). Returns
    /// false when nothing advanced.
    pub fn ack_topic_cursor(
        &self,
        py: Python<'_>,
        topic: &str,
        subscription_name: &str,
        cursor: &str,
    ) -> PyResult<bool> {
        let storage = &self.storage;
        py.detach(|| storage.ack_topic_cursor(topic, subscription_name, cursor))
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Lease up to `limit` available messages for `visibility_ms`, tracking
    /// per-message state so a nack or lease timeout redelivers just that message.
    /// Returns `(id, payload, metadata, notes, created_at)` tuples.
    pub fn lease_topic_messages(
        &self,
        py: Python<'_>,
        topic: &str,
        subscription_name: &str,
        limit: i64,
        visibility_ms: i64,
    ) -> PyResult<Vec<TopicMessageTuple>> {
        let storage = &self.storage;
        Ok(py
            .detach(|| {
                storage.lease_topic_messages(
                    topic,
                    subscription_name,
                    limit,
                    visibility_ms,
                    now_millis(),
                )
            })
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?
            .into_iter()
            .map(|m| (m.id, m.payload, m.metadata, m.notes, m.created_at))
            .collect())
    }

    /// Ack one leased message — done, never redelivered. False if nothing to ack.
    pub fn ack_message(
        &self,
        py: Python<'_>,
        topic: &str,
        subscription_name: &str,
        message_id: &str,
    ) -> PyResult<bool> {
        let storage = &self.storage;
        py.detach(|| storage.ack_message(topic, subscription_name, message_id))
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Nack one leased message — redeliver it now. False if nothing to nack.
    pub fn nack_message(
        &self,
        py: Python<'_>,
        topic: &str,
        subscription_name: &str,
        message_id: &str,
    ) -> PyResult<bool> {
        let storage = &self.storage;
        py.detach(|| storage.nack_message(topic, subscription_name, message_id))
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Per-log-subscription lag snapshot:
    /// `(topic, subscription, cursor, lag, oldest_unacked_age_ms)`.
    pub fn topic_log_stats(&self, py: Python<'_>) -> PyResult<Vec<TopicLogStatsTuple>> {
        let storage = &self.storage;
        Ok(py
            .detach(|| storage.topic_log_stats())
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?
            .into_iter()
            .map(|s| {
                (
                    s.topic,
                    s.subscription_name,
                    s.cursor,
                    s.lag,
                    s.oldest_unacked_age_ms,
                )
            })
            .collect())
    }

    /// Declare a topic (idempotent). A declared log topic retains its publishes
    /// even with no subscriber; `retention_ms` bounds a sub-less backlog.
    #[pyo3(signature = (name, mode, retention_ms=None))]
    pub fn declare_topic(
        &self,
        py: Python<'_>,
        name: &str,
        mode: &str,
        retention_ms: Option<i64>,
    ) -> PyResult<()> {
        let storage = &self.storage;
        py.detach(|| storage.declare_topic(name, mode, retention_ms))
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Every declared topic: `(name, mode, retention_ms, created_at)`.
    pub fn list_declared_topics(&self, py: Python<'_>) -> PyResult<Vec<TopicTuple>> {
        let storage = &self.storage;
        Ok(py
            .detach(|| storage.list_declared_topics())
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?
            .into_iter()
            .map(|t| (t.name, t.mode, t.retention_ms, t.created_at))
            .collect())
    }

    /// Drop ephemeral subscriptions whose owning worker is gone. Runs on the
    /// heartbeat cadence, after `reap_dead_workers` has pruned the registry.
    ///
    /// Gated behind the same reaper election as the dead-worker reap: only the
    /// leader sweeps, so the scan runs once per cluster rather than once per
    /// worker. A non-leader returns 0. `worker_id` is the election owner.
    #[pyo3(signature = (worker_id=None))]
    pub fn reap_ephemeral_subscriptions(
        &self,
        py: Python<'_>,
        worker_id: Option<&str>,
    ) -> PyResult<u64> {
        let storage = &self.storage;
        // Release the GIL: the reap scans every worker + subscription, which
        // must not freeze other Python threads while it runs.
        py.detach(|| {
            // An explicit owner elects a single reaper; without one (a manual
            // admin call) the sweep runs unconditionally.
            taskito_core::storage::sweep_ephemeral_subscriptions(storage, worker_id)
        })
        .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))
    }

    /// Publish a payload to a topic: one job per active subscription.
    /// Returns the created jobs — empty when nothing is subscribed.
    #[pyo3(signature = (topic, payload, idempotency_key=None, metadata=None, notes=None, priority=None, delay_seconds=None, max_retries=None, timeout=None, expires=None, result_ttl=None))]
    #[allow(clippy::too_many_arguments)]
    pub fn publish(
        &self,
        py: Python<'_>,
        topic: &str,
        payload: Vec<u8>,
        idempotency_key: Option<String>,
        metadata: Option<String>,
        notes: Option<String>,
        priority: Option<i32>,
        delay_seconds: Option<f64>,
        max_retries: Option<i32>,
        timeout: Option<i64>,
        expires: Option<f64>,
        result_ttl: Option<i64>,
    ) -> PyResult<Vec<PyJob>> {
        let now = now_millis();
        let scheduled_at = match delay_seconds {
            Some(d) => {
                if !d.is_finite() || d < 0.0 {
                    return Err(pyo3::exceptions::PyValueError::new_err(
                        "delay_seconds must be a finite non-negative number",
                    ));
                }
                now.checked_add((d * 1000.0) as i64).ok_or_else(|| {
                    pyo3::exceptions::PyValueError::new_err(
                        "delay_seconds too large, would overflow",
                    )
                })?
            }
            None => now,
        };
        let expires_at = match expires {
            Some(e) => {
                if !e.is_finite() || e < 0.0 {
                    return Err(pyo3::exceptions::PyValueError::new_err(
                        "expires must be a finite non-negative number",
                    ));
                }
                Some(now.checked_add((e * 1000.0) as i64).ok_or_else(|| {
                    pyo3::exceptions::PyValueError::new_err("expires too large, would overflow")
                })?)
            }
            None => None,
        };
        let result_ttl_ms = match result_ttl {
            Some(s) => Some(s.checked_mul(1000).ok_or_else(|| {
                pyo3::exceptions::PyValueError::new_err("result_ttl too large, would overflow")
            })?),
            None => None,
        };
        let timeout_ms = match timeout {
            Some(t) => Some(t.checked_mul(1000).ok_or_else(|| {
                pyo3::exceptions::PyValueError::new_err("timeout too large, would overflow")
            })?),
            None => None,
        };

        let request = PublishRequest {
            topic: topic.to_string(),
            payload,
            idempotency_key,
            metadata,
            notes,
            priority,
            scheduled_at,
            max_retries,
            timeout_ms,
            expires_at,
            result_ttl_ms,
            namespace: self.namespace.clone(),
            queue_defaults: DeliveryDefaults {
                priority: self.default_priority,
                max_retries: self.default_retry,
                timeout_ms: self.default_timeout.saturating_mul(1000),
            },
        };
        let storage = &self.storage;
        // Release the GIL: fan-out can create one job per subscription, which
        // must not block every other Python thread for the duration.
        let jobs = py
            .detach(|| publish_to_topic(storage, &request))
            .map_err(|e| pyo3::exceptions::PyRuntimeError::new_err(e.to_string()))?;
        Ok(jobs.into_iter().map(PyJob::from).collect())
    }
}
